sklearn/model_selection/_split.py

```diff
@@ -2506,6 +2508,155 @@ def _build_repr(self):
+    TRAIN: [0 1 2 3 4 5 6] TEST: [ 7  8  9 10]
+    TRAIN GROUP: ['a' 'a' 'a' 'a' 'b' 'b' 'b'] TEST GROUP: ['c' 'c' 'c' 'c']
+    TRAIN: [ 0  1  2  3  4  5  6  7  8  9 10] TEST: [11 12 13]
+    TRAIN GROUP: ['a' 'a' 'a' 'a' 'b' 'b' 'b' 'c' 'c' 'c' 'c'] TEST GROUP: ['d' 'd' 'd']
+    """
+
+    @_deprecate_positional_args
+    def __init__(self, n_splits=5, *, max_train_size=None, test_size=None, gap=0):
+        super().__init__(n_splits, shuffle=False, random_state=None)
+        self.max_train_size = max_train_size
+        self.test_size = test_size
+        self.gap = gap
+
+    def split(self, X, y=None, groups=None):
+        """Generate indices to split data into training and test set.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            Training data, where n_samples is the number of samples
+            and n_features is the number of features.
+
+        y : array-like of shape (n_samples,)
+            Always ignored, exists for compatibility.
+
+        groups : array-like of shape (n_samples,)
+            Group labels for the samples used while splitting the dataset into
+            train/test set.
+
+        Yields
+        ------
+        train : ndarray
+            The training set indices for that split.
+
+        test : ndarray
+            The testing set indices for that split.
+        """
+        if groups is None:
+            raise ValueError("The 'groups' parameter should not be None")
+        X, y, groups = indexable(X, y, groups)
+        n_samples = _num_samples(X)
+        n_splits = self.n_splits
+        n_folds = n_splits + 1
+        group_dict = {}
+        u, ind = np.unique(groups, return_index=True)
+        unique_groups = u[np.argsort(ind)]
+        n_samples = _num_samples(X)
+        n_groups = _num_samples(unique_groups)
+        for idx in np.arange(n_samples):
+            if groups[idx] in group_dict:
+                if idx - group_dict[groups[idx]][-1] == 1:
+                    group_dict[groups[idx]].append(idx)
+                else:
+                    raise ValueError(
+                        (
+                            "The groups should be continuous."
+                            " Found a non-countinuous group at"
+                            " index={0}"
+                        ).format(idx)
+                    )
+            else:
+                group_dict[groups[idx]] = [idx]
+            if n_folds > n_groups:
+                raise ValueError(
+                    (
+                        "Cannot have number of folds={0} greater than"
+                        " the number of groups={1}"
+                    ).format(n_folds, n_groups)
+                )
+        tss = TimeSeriesSplit(
+            gap=self.gap,
+            max_train_size=None,
+            n_splits=n_splits,
+            test_size=None,
+        )
+
+        for train_idx, test_idx in tss.split(unique_groups):
```

Comment:
I find the behaviour of this split quite surprising.

Here, we discard the fact that the group can be imbalanced. Therefore, we could end up with a split where the different test sets could have very different sizes.

I would have intuitively expected the following:

- apply `tss` on the original data
- find the closest split group-aligned using `group_dict`

In this case, we ensure that the test splits will have approximately the same size. However, we might have a different number of groups in the different test sets.

@getgaurav2 Did we consider this approach and if we discarded it do you recall what is the advantage of one approach compared to the other one?

Reply:
@glemaitre  - I agree with your observation  . Originally , I had thought about approaching the problem the same way as you described. 

I had realized that , after step - 1 when elements of one group will split into both Train and Test then we will need some  function to  find the "closest" split which is group aligned .  Here is one example to start  the conversation : 
- Let's say after step - 1 the split looks like : 
 TRAIN : [ 'a' , 'a' , 'a', 'a', 'a' , 'b']   TEST : ['b', 'b' , 'b' , 'b' , 'c' , 'c']
- Then after step -2  should the split be : 
 TRAIN : [ 'a' , 'a' , 'a', 'a', 'a' ]   TEST : ['b', 'b', 'b' , 'b' , 'b' , 'c' , 'c']
Reason (intuitive) : lengths of TRAIN and TEST after step -1  are equal , therefore the 'b' from TRAIN shall move into  TEST .

In my mind , This function shall be a bunch of rules that consider  -  length of splits , gap , group sizes , max_train_length , max_test_length etc. 
I could not think of an  elegant way to handle the  overlapping groups while minimizing  the difference in the length of splits.

Reply:
I don't think stratification is an appropriate objective given the other constraints here.

@glemaitre it's unclear to me what the algorithm is for finding "closest split group-aligned" that ensures each test set is distinct.